File: sweetrecycleradapter/src/main/java/com/ramo/sweetrecycleradapter/SweetRecyclerAdapter.java
package com.ramo.sweetrecycleradapter;

import android.content.Context;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class SweetRecyclerAdapter<T extends ViewTypeListener> extends RecyclerView.Adapter<SweetViewHolder<T>> {

    private Context context;
    private List<T> list;
    private List<ItemHolder> itemHolderList = new ArrayList<>();

    public SweetRecyclerAdapter(Context context, List<T> list) {
        this.context = context;
        this.list = list;
    }

    public void addHolder(int resourceId, int viewType, ISetItemView<T> iSetItemView,OnRecyclerItemClickListener<T> onRecyclerItemClickListener) {
        itemHolderList.add(new ItemHolder(resourceId, viewType, iSetItemView,onRecyclerItemClickListener));
    }

    @NonNull
    @Override
    public SweetViewHolder<T> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        for (int i = 0; i < itemHolderList.size(); i++) {
            ItemHolder<T> item = itemHolderList.get(i);
            if (viewType == item.getViewType())
                return new SweetViewHolder<>(context,parent,item);
        }
        return null;

    }

    @Override
    public void onBindViewHolder(@NonNull SweetViewHolder<T> holder, int position) {
        T item = list.get(position);
        holder.bind(item);
    }


    @Override
    public int getItemViewType(int position) {
        return list.get(position).getRecyclerViewType();
    }

    @Override
    public int getItemCount() {
        return list.size();
    }
}

File: sweetrecycleradapter/src/main/java/com/ramo/sweetrecycleradapter/SweetRecyclerAdapter.kt
package com.ramo.sweetrecycleradapter

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import java.lang.Exception
import java.util.*

class SweetRecyclerAdapter<T : ViewTypeListener>(private val list: List<T>) : RecyclerView.Adapter<SweetViewHolder<T>>() {
    private val itemHolderList: MutableList<ItemHolder<T>> = ArrayList()
    fun addHolder(resourceId: Int, viewType: Int, bindRecyclerItem: BindRecyclerItem<T>, onRecyclerItemClickListener: OnRecyclerItemClickListener<T>) {
        itemHolderList.add(ItemHolder<T>(resourceId, viewType, bindRecyclerItem, onRecyclerItemClickListener))
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SweetViewHolder<T> {
        for (i in itemHolderList.indices) {
            val item: ItemHolder<T> = itemHolderList[i]
            if (viewType == item.viewType) return SweetViewHolder(parent, item)
        }
        throw Exception("SweetRecyclerAdapter: ViewHolder is not found!")
    }

    override fun onBindViewHolder(holder: SweetViewHolder<T>, position: Int) {
        val item = list[position]
        holder.bind(item)
    }

    override fun getItemViewType(position: Int): Int {
        return list[position]!!.getRecyclerViewType()
    }

    override fun getItemCount(): Int {
        return list.size
    }
}